source/crud/crud.rst

```diff
@@ -1927,8 +1928,63 @@ The ``update`` family of operations require that the update document parameter M
 The ``replace`` family of operations require that the replacement document parameter MUST NOT begin with an atomic modifier. In practice, this means that introspection needs to happen on that document to enforce this. However, it is enough to only check the first element in the document. If it does not begin with a ``$`` sign but an element later on does, the server will throw an error.
 
 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Aggregation Pipelines with Write Stages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section discusses special considerations for aggregation pipelines that
+contain write stages (e.g. ``$out``, ``$merge``).
+
+
+Returning a cursor on the output collection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As noted in the documentation for the ``aggregate`` helper earlier in this
+document, ``$out`` and ``$merge`` are special pipeline stages that cause no
+results to be returned from the server. As such, drivers MAY setup a cursor to
+be executed upon iteration against the output collection and return that instead
+of an iterable that would otherwise have no results.
+
+
+Read preferences and server selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Historically, only primaries could execute an aggregation pipeline with ``$out``
+or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
+execute such pipelines; however, since drivers do not track
+``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
+or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
+13).
+
+Drivers MUST guarantee the following:
+
+- If an explicit (i.e. per-operation) read preference is specified for an
+  aggregation with a write stage, drivers MUST attempt to use it. If that would
+  result in a pre-5.0, secondary server being selected, drivers MUST instead
+  select a server using a primary read preference.
+- If no explicit read preference is specified but a default read preference is
+  available to inherit (e.g. from the Collection), drivers MUST attempt to use
+  it. If that would result in a pre-5.0, secondary server being selected,
+  drivers MUST instead select a server using a primary read preference.
+
+Since the server version may not be known prior to server selection, this
+specification is not prescriptive about how drivers should select a server for
+an ``aggregate`` command including a write stage. For example, drivers MAY use a
+custom server selector to consider server/wire version when matching a read
+preference and fall back to selecting a primary. Alternatively, drivers MAY
+invoke server selection an additional time if necessary.
```

Comment:
Server selection currently does not wire version into account at all. This is a complex issue that really wasn't _completely_ addressed in the driver scope. The scope states:

> Rather than requiring drivers to check out a connection from the selected server’s pool to determine its wire version, we will instead use the wire versions stored on the ServerDescription. This represents the information retrieved from the last successful SDAM check. We recognize that doing this is generally discouraged to avoid upgrade/downgrade race conditions, but actually encountering this race for this feature would require a double-downgrade of the entire cluster (5.0 to 4.2), which should be exceedingly rare. By doing this, drivers can support this feature by going through server selection once.

This citation is discussing connection check-out (assuming multi-threaded/async SDAM), but I'm more concerned with server selection at the moment. Server selection is what returns a ServerDescription, which we'd need to check to determine wire version. This made me concerned that drivers might need to invoke server selection _twice_ for this feature. The first attempt would respect a user-specified read preference and if that returns a pre-5.0 secondary we'd need to make a second selection to obtain the primary.

I asked @divjotarora about this and he clarified his original intentions while writing the scope:

> @divjotarora: IIRC, the hope was to use a "custom" server selector that is slightly more complicated than the read preference selection algorithm. I don't remember the specifics, but instead of just considering a server to be a candidate if it matches the read preference, it could be something like:
> ```
> isCandidate := (server matches readpref && server.maxWireVersion >= 13) || (server is primary)
> ```
> In the case of this being the first $out operation, you run the small risk of there being a 5.0+ secondary that hasn't been discovered yet. You could maybe handle this by removing the primary from the list of candidates if it's the only candidate and is 5.0+. You'd end up with an empty list of candidates and loop through server selection again. But regardless, I think it's possible to do the whole thing in a single 30-second selection loop rather than calling `topology.selectServer()` twice.

I don't follow the point in the second paragraph here (citing from our Slack conversation), but I think the paragraph I'm adding to the spec here is open-ended enough to allow for the pseudo-code he proposed for custom server selection.

In PHPLIB, customizing server selection in this manner isn't really an option so I think we'd need to conditionally invoke server selection a second time. We technically _could_ support it, but it would entail making non-trivial API changes to libmongoc and PHPC, which I don't think are worth the trouble in this case.

Reply:
I think the point in the second paragraph is just that a driver might still make the "wrong" choice,  If the only servers that have been discovered so far are 4.4 servers, the driver will fallback to primary, even if there are indeed 5.0 servers in the topology